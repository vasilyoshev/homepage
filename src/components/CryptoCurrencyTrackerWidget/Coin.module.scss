.coinContainer {
  display: flex;
  justify-content: center;
}

.coinRow {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-items: start;
  width: 900px;
  height: 60px;
  border-bottom: 1px solid #d7d7d7;
}

.coin {
  display: flex;
  align-items: center;
  min-width: 300px;
  padding-right: 14px;
}

.coin {
  h1 {
    width: 150px;
    font-size: 16px;
  }

  img {
    width: 30px;
    height: 30px;
    margin-right: 10px;
  }
}

.coinSymbol {
  text-transform: uppercase;
}

.coinData {
  display: flex;
  justify-content: space-between;
  width: 100%;
  padding-top: 14px;
  text-align: center;
}

.coinPrice {
  width: 110px;
}

.coinMarketcap {
  width: 200px;
}

.coinPercentRed {
  width: 60%;
  color: #f00606;
}

.coinPercentGreen {
  width: 60%;
  color: #088a08;
}
