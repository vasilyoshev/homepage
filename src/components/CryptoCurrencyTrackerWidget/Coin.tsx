import React from 'react';
import { CoinProps } from 'interfaces';
import styles from './Coin.module.scss';

export const Coin: React.FC<CoinProps> = ({ name, image, symbol, marketcap, price, priceChange }) => {
  return (
    <div className={styles.coinContainer}>
      <div className={styles.coinRow}>
        <div className={styles.coin}>
          <img src={image} alt='crypto' />
          <h1>{name}</h1>
          <p className={styles.coinSymbol}>{symbol}</p> 
        </div>
        <div className={styles.coinData}>
          <p className={styles.coinPrice}>${price.toFixed(2)}</p>       
          {priceChange < 0 ? (
            <p className={styles.coinPercentRed}>{priceChange.toFixed(2)}%</p>
          ) : (
            <p className={styles.coinPercentGreen}>{priceChange.toFixed(2)}%</p>
          )
          }
          <p className={styles.coinMarketcap}>
               Mtk Cap: ${marketcap.toLocaleString()}
          </p>
        </div>
      </div>
    </div>
  );
};
