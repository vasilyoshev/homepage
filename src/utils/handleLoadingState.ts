import axios from 'axios';
import { decrementLoadingCount, incrementLoadingCount } from 'slices';
import store from 'store';

export const handleLoadingState = (): void => {
  axios.interceptors.response.use(
    (response) => {
      setTimeout(() => store.dispatch(decrementLoadingCount()), 200);

      return response;
    },
    () => {
      store.dispatch(decrementLoadingCount());
    },
  );

  axios.interceptors.request.use((request) => {
    store.dispatch(incrementLoadingCount());

    return request;
  });
};
