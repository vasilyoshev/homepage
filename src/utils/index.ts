export * from './handleLoadingState';
export * from './testUtils';
