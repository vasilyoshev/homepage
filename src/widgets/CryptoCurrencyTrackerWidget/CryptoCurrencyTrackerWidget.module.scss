@import url('https://fonts.googleapis.com/css2?family=Montserrat&display=swap');

.coinApp {
  padding: 8px;
  font-family: 'Montserrat', sans-serif;
}

.coinText {
  margin: 0 0 16px;
  text-align: center;
}
