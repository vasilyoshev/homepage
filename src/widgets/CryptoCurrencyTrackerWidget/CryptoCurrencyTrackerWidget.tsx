import React, { ChangeEvent, useState, useEffect } from 'react';
import axios from 'axios';
import { Coin } from 'components';
import { CryptoCurrencyTrackerWidgetResProps } from 'interfaces';
import styles from './CryptoCurrencyTrackerWidget.module.scss';

export const CryptoCurrencyTrackerWidget: React.FC = () => {
  const [coinsRes, setCoinsRes] = useState<Array<CryptoCurrencyTrackerWidgetResProps>>([]);
  const [search, setSearch] = useState<string>('');

  useEffect(() => {
    function filterCoins(coinsRes: CryptoCurrencyTrackerWidgetResProps[]) {
      return coinsRes.filter((coin) => {
        return coin.name
          .toLowerCase()
          .includes(search.toLowerCase());
      });
    }
    axios.get('https://api.coingecko.com/api/v3/coins/markets?vs_currency=eur&order=market_cap_desc&per_page=5&page=1&sparkline=false')
      .then((res) => {
        setCoinsRes(filterCoins(res.data));        
      })
      .catch((err) => console.log(err));

  },[search]);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setSearch(e.target.value);
  };
  
  return (
    <div className={styles.coinApp}>
      <div className={styles.coinSearch}>
        <h1 className={styles.coinText}>Search a currency</h1>
        <form>
          <input 
            type='text'
            placeholder='Search' 
            className={styles.coinInput} 
            onChange={handleChange}
          />
        </form>
      </div> 
      {coinsRes.map((coin) => {
        return ( 
          <Coin
            key={coin.id}
            name={coin.name} 
            image={coin.image}
            symbol={coin.symbol}
            marketcap={coin.market_cap}
            price={coin.current_price}
            priceChange={coin.price_change_percentage_24h}
          />
        );
      })}
    </div>
  );
};
