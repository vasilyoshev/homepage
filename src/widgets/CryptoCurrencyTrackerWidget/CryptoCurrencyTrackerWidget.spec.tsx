import { render, screen, fireEvent } from 'utils';
import { CryptoCurrencyTrackerWidget } from './CryptoCurrencyTrackerWidget';

describe('CryptoCurrencyTrackerWidget', () => {
  it('renders successfully', () => {
    const { asFragment } = render(<CryptoCurrencyTrackerWidget />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should display currency searched by the user', async() => {
    render(<CryptoCurrencyTrackerWidget />);
    const input = screen.getByRole('textbox') as HTMLInputElement;
    fireEvent.change(input, { target: { value: 'bitcoin'}});
    expect(input.value).toBe('bitcoin');
    const element = await screen.findByText(/bitcoin/i);
    expect(element).toBeInTheDocument();
  });
});
