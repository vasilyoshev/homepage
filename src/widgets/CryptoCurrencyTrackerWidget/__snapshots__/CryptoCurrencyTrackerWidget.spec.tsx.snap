// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CryptoCurrencyTrackerWidget renders successfully 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root coinApp MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiCardContent-root coinContent"
    >
      <div
        class="coinSearch"
      >
        <h1
          class="coinText"
        >
          Crypto currency
        </h1>
        <form>
          <input
            class="coinInput"
            placeholder="search..."
            type="text"
          />
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
